rfcs/003-workspaces.md

```diff
@@ -0,0 +1,279 @@
+
+If/when the status becomes `running`, we'll try to load the inner frame.
+
+The "restart container" button should kill and relaunch the underlying workspace container. (This will be elaborated on later.)
+
+### PrairieLearn question page
+
+The "Save" button should be removed from the PrairieLearn question page, and the "Save & grade" button should be renamed to just "Grade".
+
+When the "Grade" button is clicked, the PrairieLearn web server will do one of three things:
+
+* If the workspace state is `uninitialized`, that is an invalid submission.
+* If the workspace state is `launching` or `stopped`, the PrairieLearn web server will pull the state of the workspace from S3.
+  * If that request fails, this is an error.
+* If the workspace state is `running`, the PrairieLearn web server will query the workspace host directly for its files.
+  * If that request fails or times out (this is not an error), fall back to S3.
+  * If the request to S3 fails, this is an error.
+
+Once the PrairieLearn web server has the workspace state, it will create a submission with the `gradedFiles` (specified in `info.json`) saved to `submitted_answer._files` and kick off the normal grading process. This matches the PrairieLearn convention of storing files that is used by `pl-file-upload`, `pl-file-editor`, `pl-file-preview`, etc.
+
+After the MVP, the UX could be improved to reduce the weirdness of needing to have two separate pages open at once, with editing and grading split between tabs.
+
+### Workspace hosts
+
+There will be some number of EC2 instances responsible for running the containers that power workspaces. The `workspace_hosts` table will store metadata about each host, at a minimum the hostname where it can be reached. These will be referenced by the `workspace_host_id` column in the `workspaces` table.
+
+Workspaces expose a simple API that allow them to be controlled and queried by a PrairieLearn web server. That API will be served from `/api/v1/` and include the following routes:
+
+* `POST /workspace/<workspace_id>/launch`: Begins the asynchronous process of launching a workspace container. This entails:
+  * Pull the Docker container
+  * Pull the workspace state from S3
+  * Place workspace state into a directory on disk
+  * Starting a container with the workspace state mounted as the home directory
+  * Change workspace state to `running`
+  * Emit `change:state` websocket event
+* `POST /workspace/<workspace_id>/stop`: Tears down any resources associated with this container.
+* `GET /workspace/<workspace_id>/graded_files`: Responds with a tarball including the set of graded files for this question.
```

Comment:
Or we could use WebSocket to interact with the workspace host directly, instead of POST/GET to the main server than POST/GET to the workspace host?